crates/nu-parser/src/parser.rs

```diff
@@ -222,15 +221,186 @@ pub(crate) fn check_call(
+                        // The next token is escaped so it doesn't count (only for double quote)
+                        *escaped = true;
+                    }
+                    _ => {
+                        *escaped = false;
+                    }
+                },
+            }
+        }
+
+        // Add the final span
+        match state {
+            State::Bare { from } | State::Quote { from, .. } => {
+                if from < contents.len() {
+                    spans.push(make_span(from, contents.len()));
+                }
+            }
+        }
+
+        // Log the spans that will be parsed
+        if log::log_enabled!(log::Level::Trace) {
+            let contents = spans
+                .iter()
+                .map(|span| String::from_utf8_lossy(working_set.get_span_contents(*span)))
+                .collect::<Vec<_>>();
+
+            trace!("parsing: external string, parts: {contents:?}")
+        }
+
+        // Check if the whole thing is quoted. If not, it should be a glob
+        let quoted =
+            (contents.len() >= 3 && contents.starts_with(b"$\"") && contents.ends_with(b"\""))
+                || is_quoted(contents);
+
+        // Parse each as its own string
+        let exprs: Vec<Expression> = spans
+            .into_iter()
+            .map(|span| parse_string(working_set, span))
+            .collect();
+
+        if exprs
+            .iter()
+            .all(|expr| matches!(expr.expr, Expr::String(..)))
+        {
+            // If the exprs are all strings anyway, just collapse into a single string.
+            let string = exprs
+                .into_iter()
+                .map(|expr| {
+                    let Expr::String(contents) = expr.expr else {
+                        unreachable!("already checked that this was a String")
+                    };
+                    contents
+                })
+                .collect::<String>();
+            if quoted {
+                Expression::new(working_set, Expr::String(string), span, Type::String)
+            } else {
+                Expression::new(
+                    working_set,
+                    Expr::GlobPattern(string, false),
+                    span,
+                    Type::Glob,
+                )
+            }
+        } else {
+            // Flatten any string interpolations contained with the exprs.
+            let exprs = exprs
+                .into_iter()
+                .flat_map(|expr| match expr.expr {
+                    Expr::StringInterpolation(subexprs, _) => subexprs,
+                    _ => vec![expr],
+                })
+                .collect();
+            // Make a string interpolation out of the expressions.
+            Expression::new(
+                working_set,
+                Expr::StringInterpolation(exprs, false),
+                span,
+                if quoted { Type::String } else { Type::Glob },
```

Comment:
I think this is what you meant by bareword interpolation becoming either a string or a glob type, right? I don't know if it would cause any issues, but I'd mimic how barewords are handled, just to avoid unforeseen consequences and keep things consistent. I don't know exactly how barewords operate, but my guess would be that they are turned either to `Expr::String` or `Expr::Glob`, so here we could use  `Expr::StringInterpolation`  and  `Expr::GlobInterpolation`. It's fine to add new Exprs.

Reply:
Okay, I think that's a good idea